### ISSUE TITLE
Simplify authentication and certificate issuance endpoints

### ISSUE THREAD

**Issue by codingllama** [gravitational/teleport#20343]
__Background__

Over the years Teleport authentication has undergone various changes - from single-step password+OTP login (aka ["direct login"][direct_login]), to multi-step WebAuthn ceremonies (["/mfa/" endpoints][mfa_eps]), to [various][add_mfa] [gRPC][delete_mfa] [streaming][single_use_mfa] [ceremonies][device_authn].

Some endpoints are split between [SSH][ssh_login] and [Web][web_login].

Some endpoints have [async][add_mfa] and [sync][add_mfa_sync] variants (because of lack of support for gRPC streams on Web).

SSO comes with its own, separate authentication endpoints for [GitHub][sso_gh], SAML and OIDC. (Adding WebAuthn checks to those, for example, is not out of question.)

Certificate issuance is also split between ["direct"][single_use_direct] and ["MFA"][single_use_mfa]. This often creates branching paths in clients, which have to be tested separately (and cause their own bugs when forgotten).

Authentication is ever more complex. Recent additions include [PIV support][rfd_piv] and [Device Trust][rfd_devicetrust], both which add complexity to existing ceremonies (or entirely new ceremonies, for the latter). The branching and complexity of the existing solution adds to the work necessary to improve the system, makes it difficult to exercise and reproduce scenarios, and makes it easy for entire branches to be forgotten and/or untested.

[direct_login]: https://github.com/gravitational/teleport/blob/9d97ae63fd629ddd2e42f3c617ac30f6df9d0b63/lib/client/api.go#L3347

[mfa_eps]: https://github.com/gravitational/teleport/blob/9d97ae63fd629ddd2e42f3c617ac30f6df9d0b63/lib/web/apiserver.go#L628-L635

[add_mfa]: https://github.com/gravitational/teleport/blob/9d97ae63fd629ddd2e42f3c617ac30f6df9d0b63/api/proto/teleport/legacy/client/proto/authservice.proto#L2288
[delete_mfa]: https://github.com/gravitational/teleport/blob/9d97ae63fd629ddd2e42f3c617ac30f6df9d0b63/api/proto/teleport/legacy/client/proto/authservice.proto#L2297
[single_use_mfa]: https://github.com/gravitational/teleport/blob/9d97ae63fd629ddd2e42f3c617ac30f6df9d0b63/api/proto/teleport/legacy/client/proto/authservice.proto#L2076
[device_authn]: https://github.com/gravitational/teleport/blob/9d97ae63fd629ddd2e42f3c617ac30f6df9d0b63/api/proto/teleport/devicetrust/v1/devicetrust_service.proto#L109

[ssh_login]: https://github.com/gravitational/teleport/blob/9d97ae63fd629ddd2e42f3c617ac30f6df9d0b63/lib/web/apiserver.go#L630
[web_login]: https://github.com/gravitational/teleport/blob/9d97ae63fd629ddd2e42f3c617ac30f6df9d0b63/lib/web/apiserver.go#L631

[add_mfa_sync]: https://github.com/gravitational/teleport/blob/9d97ae63fd629ddd2e42f3c617ac30f6df9d0b63/api/proto/teleport/legacy/client/proto/authservice.proto#L2299

[sso_gh]: https://github.com/gravitational/teleport/blob/9d97ae63fd629ddd2e42f3c617ac30f6df9d0b63/lib/web/apiserver.go#L624-L626

[single_use_direct]: https://github.com/gravitational/teleport/blob/9d97ae63fd629ddd2e42f3c617ac30f6df9d0b63/api/proto/teleport/legacy/client/proto/authservice.proto#L2071

[rfd_piv]: https://github.com/gravitational/teleport/blob/master/rfd/0080-hardware-key-support.md
[rfd_devicetrust]: https://github.com/gravitational/teleport.e/blob/master/rfd/0001e-device-trust.md

__Objective__

The objective is simple: unify authentication in a single logical set of endpoints.

It is best to use synchronous RPCs for wider compatibility.

The ceremony has to be capable of multiple steps, and it has to be capable of being gracefully extended in the future (recent additions include U2F/WebAuthn challenges, Passwordless, PIV attestation and Device Trust challenges).

An additional plus is the capability to perform the ceremony purely using gRPC, meaning an initial, unauthenticated connection to either Proxy (similar to existing ones) or Auth. (And yes, Web will need a proper gRPC client for this.)

Certificate issuance often includes some sort of re-authentication and could use a similar solution.

Various other actions would like to be gated by challenges as well (eg, require_session_mfa and [MFA for administrative actions](https://github.com/gravitational/teleport/issues/9089)). [CreateAuthenticateChallenge](https://github.com/gravitational/teleport/blob/9d97ae63fd629ddd2e42f3c617ac30f6df9d0b63/api/proto/teleport/legacy/client/proto/authservice.proto#L2307) does a reasonable job of it, but we could lay out a definitive approach for re-authentication as well.

**Comment by codingllama** [comment]
aka Uber Auth.

**Comment by codingllama** [comment]
Other endpoints could also do with a unified API (eg, creating and managing various tokens), but it's fair to say that these are out of scope here.

**Comment by codingllama** [comment]
I've got about a month to hack on this, so instead of a design->review->implementation approach, which wouldn't complete, I'm breaking this into a series of small-ish refactors:

1. Address stream vs unary RPC duplication

    Main culprits here are:

    - AddMFADevice{,Sync}
    - DeleteDeviceMFA{,Sync}
    - GenerateUserCerts vs GenerateUserSingleUseCerts

    The GenerateUserCerts family needs a closer look, as I don't think those are necessarily functionally equivalent.

    This moves us closer to a world where CreateAuthenticateChallenge and CreateRegisterChallenge have a larger part, makes tsh and Web UI use similar logic, and possibly simplifies user cert issuance.

    Lets us eventually delete all streaming MFA and issuance ceremonies.

2. Unify SSH and Web logins

    Likely done exclusively in the Auth layer. Web still has a needlessly large endpoint surface but at least this makes an unified "AuthenticateUser" RPC simpler to write.

    Lets us eventually delete a bunch of web endpoints and a few methods in Auth.

3. Move various of the aforementioned RPCs to a new AuthenticationService

    Main targets here are:

    - CreateAuthenticateChallenge
    - CreateRegisterChallenge
    - GenerateUserCerts / GenerateUserSingleUseCerts
    - SSH / Web logins

    This lets us break away some logic from the monolithic "AuthService", puts authentication logic in the same umbrella and lets clients use the same RPCs for both unauthenticated and authenticated use-cases.

    Lets us eventually delete a huge chunk of web endpoints and funnel tsh into easy to predict paths.

All we complete would land on v15, barring the need to backport because of subsequent changes, meaning a code cleanup could land on v17 (one major of leeway for tsh).

**Comment by codingllama** [comment]
## Uber Auth Mk1 tracker

```[tasklist]
### Unary vs Stream
- [ ] AddMFADevice
- [ ] DeleteMFADevice
- [ ] GenerateUserCerts vs GenerateUserSingleUseCerts
```

```[tasklist]
### Unify logins
- [ ] SSH vs Web
```

```[tasklist]
### AuthenticationService (stretch)
- [ ] Service definitions
- [ ] Unauthenticated service available (distinc port)
- [ ] Authenticated service available (same port as Auth)
- [ ] CreateAuthenticateChallenge
- [ ] CreateRegisterChallenge
- [ ] AuthenticateUser
- [ ] GenerateUserCerts
- [ ] Web UI uses (unauthenticated) AuthenticationService
- [ ] Web UI endpoints deprecated
```

```[tasklist]
### Concurrent device logins (stretch)
- [ ] Device challenges issued via CreateAuthenticateChallenge
- [ ] Device authentication happens as part of AuthenticateUser
- [ ] tsh performs ceremonies "concurrently"
```

**Comment by codingllama** [comment]
## Uber Auth Mk1 tracker

### Unary vs Stream
- [ ] AddMFADevice
- [ ] DeleteMFADevice
- [ ] GenerateUserCerts vs GenerateUserSingleUseCerts

### Unify logins
- [ ] SSH vs Web

### AuthenticationService (stretch)
- [ ] Service definitions
- [ ] Unauthenticated service available (distinc port)
- [ ] Authenticated service available (same port as Auth)
- [ ] CreateAuthenticateChallenge
- [ ] CreateRegisterChallenge
- [ ] AuthenticateUser
- [ ] GenerateUserCerts
- [ ] Web UI uses (unauthenticated) AuthenticationService
- [ ] Web UI endpoints deprecated

### Concurrent device logins (stretch)
- [ ] Device challenges issued via CreateAuthenticateChallenge
- [ ] Device authentication happens as part of AuthenticateUser
- [ ] tsh performs ceremonies "concurrently"

All cleanups are slotted for after v16, as stated above.

**Comment by Tener** [comment]
This is a very welcome stream of work, thank you for tackling that. 

What do you think about having an end-to-end review of the changes before the v15 release or after this is finished?

**Comment by codingllama** [comment]
> This is a very welcome stream of work. Thank you for tackling this!

Thanks! :D

> What do you think about having an end-to-end review of the changes before the v15 release or after this is finished?

Sure, that would be nice. Are you volunteering? ;)

**Comment by Tener** [comment]
> Sure, that would be nice. Are you volunteering? ;)

Hah, why not, feel free to ping me :-)

**Comment by codingllama** [comment]
Just a quick update, but I think short-term this has seen most of the work it'll get. There are some cleanups for post v15, everything else is slotted to some future date.
